File: python_port/Basin_Pendulum/Basin_test.py
import numpy as np
import matplotlib.pyplot as plt
import Pen_stable_check

# Assuming the is_stable function from the previous example is defined

#warning, this test runs exceedingly slow

n = 200 #originally 200
theta_range = np.linspace(-np.pi, np.pi, n)
omega_range = np.linspace(-20, 20, n)

# Initialize results array
results = np.zeros((n, n), dtype=int)

print("Evaluating stable points...")
# Evaluate stability for each combination of initial angle and angular velocity
for i, theta_0 in enumerate(theta_range):
    for j, omega_0 in enumerate(omega_range):
        results[i, j] = Pen_stable_check.is_stable(theta_0, omega_0)
print("stable points evaluated")

# Find the stable (1) points
stable_points = np.argwhere(results == 1)

# Convert index to angle and angular velocity
stable_points_xy = [(theta_range[i], omega_range[j]) for (i, j) in stable_points]

# Unzip stable points for plotting
stable_theta, stable_omega = zip(*stable_points_xy)

# Plotting
plt.figure(figsize=(10, 6))
plt.scatter(stable_theta, stable_omega, marker='o', color='green', s=2, label="Stable")
plt.xlabel("Initial Angle (rad)")
plt.ylabel("Initial Angular Velocity (rad/s)")
plt.xlim(-np.pi, np.pi)
plt.ylim(-20, 20)
plt.title("Stable Points of the Pendulum")
plt.legend()
plt.grid()
plt.show()

File: python_port/Basin_Pendulum/Basin_test.jl
using Plots

include("Pen_stable_check.jl")

n = 200
θ_range = range(-pi, pi, length=n)
ω_range = range(-20, 20, length=n)

results = zeros(Int, n, n)

for (i, θ₀) in enumerate(θ_range)
    for (j, ω₀) in enumerate(ω_range)
        results[i, j] = is_stable(θ₀, ω₀)
    end
end

# Find the stable (1) points
stable_points = [(i, j) for i in 1:n, j in 1:n if results[i, j] == 1]

# Convert index to angle and angular velocity
stable_points_xy = [(θ_range[i], ω_range[j]) for (i, j) in stable_points]


display(scatter(stable_points_xy, markershape=:circle, markercolor=:green, markerstrokewidth=0, 
label="Stable", xlabel="Initial Angle (rad)", ylabel="Initial Angular Velocity (rad/s)", 
xlims=(-pi, pi), ylims=(-20, 20), markersize=2))
